[16.0][MIG] account_payment_tier_validation: backport to v16

quirino95 opened the pull request.

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
```diff
@@ -1,14 +1,42 @@
 
     _tier_validation_manual_config = False
+
+    def action_post(self):
```
This method is overwriting core's method defined in https://github.com/odoo/odoo/blob/b0be89a67d0ce458db99f4331e21ed56d3c4774a/addons/account/models/account_payment.py#L934 without calling it with `super`, this breaks the inheritance chain.
Please add a call to `super`'s method.

SirPyTech replied:
👍

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
```diff
@@ -1,14 +1,42 @@
+                        "the recipient bank account must be manually validated. "
+                        "You should go on the partner bank account of %(partner)s"
+                        "in order to validate it.",
+                        method_name=self.payment_method_line_id.name,
```
```suggestion
                        method_name=payment.payment_method_line_id.name,
```

SirPyTech replied:
:+1:

SirPyTech reviewed (changes requested):
Thanks for the PR!
Please check the migration process that is used in OCA for migrating to this major version: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0.
This can be adapted to a backward-migration (like this one, from `18.0` to `16.0`) adapting a few steps, for instance
```sh
git format-patch --keep-subject --stdout origin/16.0..origin/15.0 -- $module | git am -3 --keep
```
in this case should be
```sh
git format-patch --keep-subject --stdout origin/16.0..origin/18.0 -- $module | git am -3 --keep
```
Also note that commits should be squashed according to https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests.
Right now the commits are 
<img width="1485" height="510" alt="image" src="https://github.com/user-attachments/assets/4bad785e-c60b-42e6-918c-cda5845ef966" />
and some of them should be squashed together.

---
Please fix the tests because right now they are failing: 
<img width="1073" height="543" alt="image" src="https://github.com/user-attachments/assets/98022574-5a9f-4470-8680-c95af681c53e" />
If the CI fails then this PR can't be merged.
The first failure is:
```
2025-07-18 12:37:19,765 330 ERROR odoo odoo.addons.account_payment_line.tests.test_account_payment_line: FAIL: TestAccountPaymentLines.test_01_customer_payment
Traceback (most recent call last):
  File "/__w/account-payment/account-payment/account_payment_line/tests/test_account_payment_line.py", line 201, in test_01_customer_payment
    self.assertTrue(new_payment.is_reconciled)
AssertionError: False is not true
```
(from https://github.com/OCA/account-payment/actions/runs/16370675214/job/46258021642?pr=859#step:8:273)
I know it's not the module you are migrating, but if this PR is causing it to fail you have two options:
- (best) adapt the code so that it doesn't cause a failure in other modules
- mark both modules as not compatible and split the CI so that they are tested independently.
  See https://github.com/OCA/oca-addons-repo-template/blob/d1483d585edc2323e8faba77c9294e5e2a4176b8/copier.yml#L147-L153:
  >Are there in this repo modules that don't get along with their friends? If so, list them here (YAML format) and they will be tested in separate jobs.
  >Beware, if rebel modules should stay separated in groups, you should join them with commas, which could be misinterpreted by YAML.
  >Example: ["rebel_module_1,rebel_module_2", even_more_rebel_module]

---
You might want to look at related implementations like:
- **account_payment_order_tier_validation** for `16.0` https://github.com/OCA/bank-payment/tree/4432cddb0c747298be581f71ec883197c19c7c94/account_payment_order_tier_validation
- a previous attempt of implementing this same module: https://github.com/OCA/account-payment/pull/664

quirino95 commented:
Thank you @SirPyTech !
I did `format-patch` and squashed commits. I've updated also behavior in `account_payment.py` by taking code from the "previous attempt", and as consequence all tests are now ok!

SirPyTech commented on account_payment_tier_validation/models/account_move.py:
**question**: I don't think this file is needed in `16.0` (maybe not even in `18.0`), could you please check and remove it eventually?

quirino95 replied:
Confirm, it's not needed. Removed.

SirPyTech replied:
:+1:

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
**chore**: This file (and most of the module) is an almost exact copy of https://github.com/OCA/account-payment/pull/664/files#diff-9cef5e44f9cd3a54a978109fc6e89d61feb1d100206e820022a24efbf5e2a3ca, please at least add the author @alexeirivera87 as a co-author (see https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors) of the commit.
And I dare say they also deserve an author spot in the module itself.

quirino95 replied:
Of course, I've added Alexei as co-author in my last commit.

SirPyTech replied:
:+1:

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
```diff
@@ -0,0 +1,78 @@
+    _inherit = ["account.payment", "tier.validation"]
+    _state_from = ["draft"]
+    _state_to = ["posted"]
+    _cancel_state = "canceled"
```
**suggestion**: This is already declared in `tier.validation` (https://github.com/OCA/server-ux/blob/a6a3fb51c0f85f2031da1aba09328c1deb3b2fb4/base_tier_validation/models/tier_validation.py#L34), is it really needed here?

SirPyTech replied:
:+1:

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
```diff
@@ -0,0 +1,78 @@
+class AccountPayment(models.Model):
+    _name = "account.payment"
+    _inherit = ["account.payment", "tier.validation"]
+    _state_from = ["draft"]
```
**suggestion**: This is already declared in `tier.validation` (https://github.com/OCA/server-ux/blob/a6a3fb51c0f85f2031da1aba09328c1deb3b2fb4/base_tier_validation/models/tier_validation.py#L32), is it really needed here?

SirPyTech replied:
:+1:

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
```diff
@@ -0,0 +1,78 @@
+    _cancel_state = "canceled"
+
+    _tier_validation_manual_config = False
+
```
**question**: comparing this file with https://github.com/OCA/account-payment/pull/664/files#diff-9cef5e44f9cd3a54a978109fc6e89d61feb1d100206e820022a24efbf5e2a3ca I see here we are missing the override of `_get_under_validation_exceptions`, was it causing any issues? Otherwise I'd say we can keep it.

quirino95 replied:
No issues keeping that method.

SirPyTech replied:
:+1:

SirPyTech commented on account_payment_tier_validation/models/account_payment.py:
```diff
@@ -0,0 +1,78 @@
+        """
+        self.ensure_one()
+        if self._context.get("validate_in_progress"):
+            return True
+        return False
+
+    def _get_need_validation(self):
+        """
+        Needed because state is a related field not stored,
+        _compute_need_validation (field need_validation) in model
+        tier_validation, does not properly work when executing button
+        "Validate"
+        """
+        self.ensure_one()
+        if isinstance(self.id, models.NewId):
+            return False
+        tiers = self.env["tier.definition"].search([("model", "=", self._name)])
+        valid_tiers = any([self.evaluate_tier(tier) for tier in tiers])
+        return (
+            not self.review_ids
+            and valid_tiers
+            and self._context.get("validate_in_progress")
+        )
```
**suggestion**: This is a slightly modified copy (the only change being the `context` usage) of an old version of `_compute_need_validation` (https://github.com/OCA/server-ux/blob/abf06d7e9365080987a2181205cbf017da736671/base_tier_validation/models/tier_validation.py#L216-L225):
```py
    def _compute_need_validation(self):
        for rec in self:
            if isinstance(rec.id, models.NewId):
                rec.need_validation = False
                continue
            tiers = self.env["tier.definition"].search([("model", "=", self._name)])
            valid_tiers = any([rec.evaluate_tier(tier) for tier in tiers])
            rec.need_validation = (
                not rec.review_ids and valid_tiers and rec._check_state_from_condition()
            )
```

Right now the code has been improved (https://github.com/OCA/server-ux/blob/a6a3fb51c0f85f2031da1aba09328c1deb3b2fb4/base_tier_validation/models/tier_validation.py#L251-L271):
```py
    def _compute_need_validation(self):
        for rec in self:
            if isinstance(rec.id, models.NewId):
                rec.need_validation = False
                continue
            tiers = (
                self.env["tier.definition"]
                .with_context(active_test=True)
                .search(
                    [
                        ("model", "=", self._name),
                        ("company_id", "in", [False] + rec._get_company().ids),
                    ]
                )
            )
            valid_tiers = tiers.filtered(lambda x: rec.evaluate_tier(x))
            requested_tiers = rec.review_ids.filtered(
                lambda x: x.status != "pending"
            ).mapped("definition_id")
            new_tiers = valid_tiers - requested_tiers
            rec.need_validation = new_tiers and rec._check_state_from_condition()
```
, for instance with multi-company management, could you please update this code accordingly?

Or even better, we could try adding a `depends_context` to an override of `_compute_need_validation` so that we won't need to keep this copy updated, what do you think?

quirino95 replied:
I went with the update of the code, after several attempts to get override to work. As mentioned in docstring, `_compute_need_validation` in this case doesn't work properly, so I preferred to keep `_get_need_validation`, improved with multi-company management.

SirPyTech replied:
That's fine, I understand it's not easy to adapt an override to work with the new context value.

But what about the last few lines of the above method?
```py
            valid_tiers = tiers.filtered(lambda x: rec.evaluate_tier(x))
            requested_tiers = rec.review_ids.filtered(
                lambda x: x.status != "pending"
            ).mapped("definition_id")
            new_tiers = valid_tiers - requested_tiers
            rec.need_validation = new_tiers and rec._check_state_from_condition()
```
 I see they are still missing in the latest code of this PR:
```py
        valid_tiers = tiers.filtered(lambda x: self.evaluate_tier(x))
        return (
            not self.review_ids
            and valid_tiers
            and self._context.get("validate_in_progress")
        )

```

SirPyTech replied:
Here your reply from https://github.com/OCA/account-payment/pull/859#discussion_r2236240293:
> using the "new code", behavior changes and method doesn't work anymore as expected. No ValidationError is throwed so validation is always ok.
Looking at server-ux module, i saw that status field in tier.review in 18.0 has a new waiting selection, missing in 16.0. I did some attempts to fix behavior with new code but without success.
Anyway, all tests are ok and the behavior is the expected one, exactly as in v18.

What I am proposing has nothing to do with `18.0`: the additional code has been added in `16.0` (where there is no `waiting` status`) with https://github.com/OCA/server-ux/commit/a64e79194efdef7589798131f0930ab751aaccba, in the commit itself and its PR https://github.com/OCA/server-ux/pull/928 you can find the reason behind it.

From a more generic point of view: the implementation in this model is kind of a _needed hack_ for tier validation to work in a model where `state` field is related, so some of the usual methods in `tier.validation` do not get triggered properly.
The _needed hack_ consists of using the context `validate_in_progress` to recognize when we need to validate the record: this is then used in places like `_check_state_conditions_custom`, while in `tier.validation` it is done in `_check_state_conditions` and `_check_state_from_condition`.
So this model has to re-implement all the places where such methods are used and:
- if there is `_check_state_from_condition` then use the context `validate_in_progress`
- if there is `_check_state_conditions` then use `_check_state_conditions_custom` instead

This has been done where needed: `write` and `need_validation` methods (see https://github.com/OCA/server-ux/blob/23e96d195dd78187789f42264a206350e470b01c/base_tier_validation/models/tier_validation.py).
As I suggested in https://github.com/OCA/account-payment/pull/859#discussion_r2224840767, doing this with overrides would be more maintainable, but since it's technically harder, it's ok to copy/paste the methods.

If the copy-paste is not in sync (like in this piece of code), that would cause an unexpected behavior because payments validation would behave differently than any other tier-validated model (like invoices, when **account_move_tier_validation** is installed).

SirPyTech replied:
:+1:

SirPyTech reviewed (changes requested):
**chore**: Please check again the commit history, right now it's
<img width="1532" height="717" alt="Image" src="https://github.com/user-attachments/assets/4021718e-61b4-40fd-8d9b-29c50830bcfd" />
but according to https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0 it should be:

1. All commits that come from `git format-patch` (mentioned in https://github.com/OCA/account-payment/pull/859#pullrequestreview-3037059281), merged according to https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate
2. An automatic commit generated by `pre-commit`
3. The `[MIG]` commit with the needed manual changes to make the module work in `16.0`

Right now there are some commits after the `[MIG]` one that shouldn't be there.

---

@alexeirivera87 we are reviving your old PR #664, would you like to have a look?

SirPyTech commented on account_payment_tier_validation/tests/test_tier_validation.py:
```diff
@@ -57,9 +57,7 @@ def test_02_validation_account_payment(self):
-        payment.amount = 100
-        payment.action_post()
-        self.assertEqual(payment.state, "in_process")
+        self.assertEqual(payment.state, "draft")
```
**question**: Looks like the `in_process` status of `18.0` corresponds to the `posted` status in `16.0`, so why is `draft` used here instead?

quirino95 replied:
Updated test using `posted` instead of `draft` (added also the two lines before in order to don't break tests.

quirino95 replied:
Can't reply to previous thread (`_get_need_validation`), so wrote response here: using the "new code", behavior changes and method doesn't work anymore as expected. No ValidationError is throwed so validation is always ok.
Looking at `server-ux` module, i saw that `status` field in `tier.review` in 18.0 has a new `waiting` selection, missing in 16.0. I did some attempts to fix behavior with new code but without success.
Anyway, all tests are ok and the behavior is the expected one, exactly as in v18.

SirPyTech replied:
> Can't reply to previous thread (`_get_need_validation`)

The thread is https://github.com/OCA/account-payment/pull/859#discussion_r2224840767

SirPyTech replied:
> Updated test using `posted` instead of `draft` (added also the two lines before in order to don't break tests.

:+1:

quirino95 commented:
Thank you Simone for the explanation. Now it's clearer for me, so I updated code.
I hope now everything is fine.

quirino95 commented:
After https://github.com/OCA/server-ux/pull/1135, I had to change a couple of occurences of `validated` field, which not exists in `server-ux` for `16.0`.  I replaced it with `validation_status = 'validated'`

rvalyi commented:
/ocabot merge nobump

SirPyTech commented:
Rebased to check merge failure

quirino95 commented:
Fixed tests failure!

SirPyTech commented:
> /ocabot merge nobump

(from https://github.com/OCA/account-payment/pull/859#issuecomment-3592697369)

@rvalyi please try again

alexeirivera87 commented:
@SirPyTech  @quirino95 

Thanks for considering my old PR

rvalyi commented:
/ocabot merge nobump

marcos-mendez commented:
Hi @OCA-git-bot, thank you for taking the time to respond to our automated review! Your feedback is valuable and helps us improve.

We've logged this as an issue to review and improve our process: https://git.pop.coop/pop/oca-reviewer/issues/2

We'll work on addressing the points you raised.

---
_Auto-reply by [oca-feedback](https://git.pop.coop/pop/oca-feedback)_

SirPyTech commented:
> /ocabot merge nobump

Thanks you very much @rvalyi for trying to merge (ignoring AI review), the merge failed for the known issue mentioned in https://github.com/OCA/oca-addons-repo-template/pull/333, we only have to update the repository template.

SirPyTech commented:
@marcos-mendez please dismiss the generated review, this PR was "ready to merge" before that

SirPyTech commented:
> > /ocabot merge nobump
> 
> Thanks you very much @rvalyi for trying to merge (ignoring AI review), the merge failed for the known issue mentioned in [OCA/oca-addons-repo-template#333](https://github.com/OCA/oca-addons-repo-template/pull/333), we only have to update the repository template.

We should be fine because https://github.com/OCA/account-payment/pull/954 has been merged, could you please retry?
